### PR TITLE
Add binary exponential back-off algorithm

### PR DESCRIPTION
**Pull request by baierjan**
Add option to enable back-off algorithm which will double the sleep amount between each retry. I can drop the commit with tests if desired (as the length of the test is now significantly larger).

### CODE REVIEW

File: retry [hunk 1]
```diff
@@ -6,24 +6,28 @@ options:
     -h,--help               Show this help
     -r,--retries=RETRIES    How many retries to do on command failure after the
                             initial try. Defaults to 3.
-    -s,--sleep=SLEEP        How many seconds to sleep between retries
+    -s,--sleep=SLEEP        How many seconds to sleep between retries. Defaults
+                            to 3 seconds.
+    -e,--exponential        Enable simple binary exponential back-off
+                            algorithm. Disabled by default.
```
Comment (okurz) [inline retry]: Awesome! I already thought about exponential back-off and you already implement it! :) Here I had the idea to not have a boolean variable but a factor defaulting to 1 and multiplying the sleep time. How about

```
-S,--sleep-factor=SLEEP_FACTOR …something about exponential back-off… Defaults to 1 meaning same sleep time between each try.
```

File: retry [hunk 2]
```diff
@@ -6,24 +6,28 @@ options:
...
   case "$1" in
-    -h | --help    ) usage 0; shift   ;;
-    -r | --retries ) retries=$2; shift 2 ;;
-    -s | --sleep   ) sleep=$2; shift 2 ;;
-    --             ) shift; break ;;
-    *              ) break ;;
+    -h | --help        ) usage 0; shift ;;
+    -r | --retries     ) retries=$2; shift 2 ;;
+    -s | --sleep       ) sleep=$2; shift 2 ;;
+    -e | --exponential ) exponential=2; shift ;;
```
Comment (okurz) [inline retry]: … aligned with what I suggested above then this would just read out `$2`

File: retry [hunk 3]
```diff
@@ -35,5 +39,6 @@ until "$@"; do
     echo "Retrying up to $retries more times after sleeping ${sleep}s …" >&2
     retries=$((retries-1))
     sleep "$sleep"
+    [ -n "$exponential" ] && sleep=$((sleep*exponential))
```
Comment (okurz) [inline retry:43]: aligned with the above suggestion then simply

```suggestion
    sleep $((sleep*sleep_factor))
```

File: test/01-retry.t [hunk 4]
```diff
@@ -28,5 +28,8 @@ is "$(retry true)" '' 'successful command does not show any output by default'
 set +e; out=$(retry -s 0 false 2>&1); rc=$?; set -e
 like "$out" 'Retrying up to 3 more.*Retrying up to 1' 'failing command retries'
 is $rc 1 'failing command returns no success'
+set +e; out=$(retry -s 1 -e -r 2 false 2>&1); rc=$?; set -e
+like "$out" 'sleeping 1s.*sleeping 2s' 'sleep ammount doubles'
+is $rc 1 'failing command returns no success'
```
Comment (okurz) [inline test/01-retry.t]: so above I was using simply `-s 0` to kinda "mock" the sleeping but I am sure we can find a better approach that overrides the actual sleep implementation. Maybe https://github.com/os-autoinst/scripts/blob/master/test/01-label-known-issues.t#L21 gives an idea how to mock sleep but check what argument we actually calculated without actually sleeping?

**Review by okurz — CHANGES_REQUESTED** [review]

File: test/01-retry.t (hunk 4 above)
Comment (baierjan) [reply test/01-retry.t]: That is a good idea, I will try to implement that.

File: retry (hunk 1 above)
Comment (baierjan) [reply retry]: I experimented with using the SLEEP as the base factor, but that did not feel good enough for larger sleep values so I decided to use the binary base and multiply that with SLEEP.

File: retry (hunk 3 above)
Comment (baierjan) [reply retry:43]: I still need to remember something between the cycles otherwise we only get a simple constant multiplication.

File: retry (hunk 2 above)
Comment (baierjan) [reply retry]: Yes, I was thinking if I should make the base configurable and decided for simplicity.. As you requested more features, I will extend it.

**Comment by baierjan** [comment]
What about now? You can use `retry -e cmd` to use the binary one (i.e. double the sleep after each iteration: 3,6,12), or you can tweak the base factor as in `retry -e10 -s1 cmd` or `retry --exponential=10 --sleep=1 cmd` to obtain 1,10,100 sleep sequence.

File: test/01-retry.t [hunk 5]
```diff
@@ -25,8 +28,11 @@ is $rc 1 'calling with unknown option returns failure'
 like "$out" 'unrecognized option.*usage:' 'retry help is shown for unknown option'
 ok "$(retry true)" 'successful command returns success'
 is "$(retry true)" '' 'successful command does not show any output by default'
-set +e; out=$(retry -s 0 false 2>&1); rc=$?; set -e
+set +e; out=$(retry false 2>&1); rc=$?; set -e
 like "$out" 'Retrying up to 3 more.*Retrying up to 1' 'failing command retries'
 is $rc 1 'failing command returns no success'
-set +e; out=$(retry -s 0 -r 1 -- sh -c 'echo -n .; false' 2>/dev/null); set -e
+set +e; out=$(retry -s 1 -e -r 2 false 2>&1); rc=$?; set -e
+like "$out" 'sleeping 1s.*sleeping 2s' 'sleep ammount doubles'
```
Comment (okurz) [inline test/01-retry.t]:
```suggestion
like "$out" 'sleeping 1s.*sleeping 2s' 'sleep amount doubles'
```

**Review by okurz — CHANGES_REQUESTED** [review]

File: test/01-retry.t (hunk 5 above)
Comment (baierjan) [reply test/01-retry.t]: Ups... right.

**Review by okurz — APPROVED** [review]

**Pull request merged**